content/multi-threading.md

```diff
@@ -0,0 +1,381 @@
+to the dbus tree can only occur after the method has completed. This can
+be observed by a client, if the client invokes a second method immediately
+after the first has returned. For example, if the client invokes the CreatePool
+method and then immediately invokes the GetManagedObjects() method, some
+pool object paths corresponding to the pool or its devices may not yet
+be present in the tree. The opposite behavior can also be observed, for
+example, if the client invokes the DestroyPool method, some object paths
+belonging to the destroyed pool may still be found by a GetManagedObjects()
+invocation.
+
+2. If two D-Bus methods are invoked in separate processes, the same behaviors
+described in (1) are somewhat easier to observe.
+
+3. We believe that we have made it impossible to incorrectly update the
+tree by returning rich result types from the engine methods.
+
+Given two distinct mutating D-Bus methods running in separate threads
+there is a possibility of a situation rather analogous to a race-condition
+arising. Two tasks may read the dbus tree, update the internal engine state,
+and then send update messages on the dbus channel. It is uncertain which
+task will acquire the engine mutex. This is
+partially analogous to the classic race-condition where two processes read a
+single variable, and then both update that variable in an undetermined order.
+
+What makes this analogy only partial is the interposition of the engine,
+which restricts the updates that may be requested of the DbusTreeHandler by
+the DbusRequestHandler. The engine methods invoked by the D-Bus layer return
+a result which sufficiently distinguishes the actions actually taken by
+the engine so that conflicting updates to the dbus tree can not be
+requested. Thus the updates are constrained to be correct.
+
+For example, consider that two conflicting commands may be handled at the
+same time: one command to delete a filesystem and the other to rename
+the same filesystem. If both commands are being handled in separate threads
+each will read the same data based on the filesystem object. Then, either
+one may enter the engine mutex. If the rename task enters the mutex first,
+it will be the first to place a message on the dbus channel. The DbusTreeHandler
+will remove the rename message first and then the remove message placed
+on the dbus channel after the remove request completes. Clearly,
+this order of processing can not result in an error. With the other order,
+the remove message will be placed on the dbus channel before the rename
+occurs. But in this case, the engine method will return a result indicating
+that no rename could occur, because the filesystem could not be found.
+Consequently, no rename message will be put on the dbus channel, and so the
+DbusTreeHandler will receive the remove message only. Thus, no incorrect
+update is performed on the dbus tree.
+
+Error Behavior
+--------------
+`stratisd` exits immediately if any task encounters an error.
```

Comment:
It may be worth clarifying this as what is really going on in that case is that it exists if event processing on any of the streams begins to fail or the streams are closed.

Reply:
I've made the exiting discussion considerably longer.